patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs

```diff
@@ -0,0 +1,426 @@
+				if (customLoc) {
+					xLoc = mAccSlot.XLoc;
+					yLoc = mAccSlot.YLoc;
+				}
+				else if (!SetDrawLocation(num20, slot + Player.dye.Length - 3, skip, ref xLoc, ref yLoc))
+					return true;
+
+				var thisSlot = Get(slot);
+
+				if (thisSlot.DrawFunctionalSlot) 
+					DrawFunctional(ModSlotPlayer(Player).exAccessorySlot, ModSlotPlayer(Player).exHideAccessory, -10, slot, flag3, xLoc, yLoc);
+				if (thisSlot.DrawVanitySlot)
+					DrawSlot(ModSlotPlayer(Player).exAccessorySlot, -11, slot + list.Count, flag3, xLoc, yLoc);
+				if (thisSlot.DrawDyeSlot)
+					DrawSlot(ModSlotPlayer(Player).exDyesAccessory, -12, slot, flag3, xLoc, yLoc);
+			}
+			else {
+				if (!SetDrawLocation(num20, slot - 3, skip, ref xLoc, ref yLoc))
+					return true;
+
+				DrawFunctional(Player.armor, Player.hideVisibleAccessory, 10, slot, flag3, xLoc, yLoc);
+				DrawSlot(Player.armor, 11, slot + Player.dye.Length, flag3, xLoc, yLoc);
+				DrawSlot(Player.dye, 12, slot, flag3, xLoc, yLoc);
+			}
+
+			return !customLoc;
+		}
+
+		/// <summary>
+		/// Applies Xloc and Yloc data for the slot, based on ModAccessorySlotPlayer.scrollSlots
+		/// </summary>
+		internal bool SetDrawLocation(int num20, int trueSlot, int skip, ref int xLoc, ref int yLoc) {
+			int accessoryPerColumn = GetAccessorySlotPerColumn(num20);
+			int xColumn = (int)(trueSlot / accessoryPerColumn);
+			int yRow = trueSlot % accessoryPerColumn;
+						
+			if (ModSlotPlayer(Player).scrollSlots) {
+
+				int row = yRow + (xColumn) * accessoryPerColumn - ModSlotPlayer(Player).scrollbarSlotPosition - skip;
+
+				yLoc = (int)((float)(num20) + (float)((row + 3) * 56) * Main.inventoryScale) + 4;
+				int chkMin = (int)((float)(num20) + (float)((0 + 3) * 56) * Main.inventoryScale) + 4;
+				int chkMax = (int)((float)(num20) + (float)(((accessoryPerColumn - 1) + 3) * 56) * Main.inventoryScale) + 4;
+
+				if (yLoc > chkMax || yLoc < chkMin) {
+					return false;
+				}
+
+				xLoc = Main.screenWidth - 64 - 28;
+			}
+
+			else {
+				int row = yRow, tempSlot = trueSlot, col = xColumn;
+				if (skip > 0) {
+					tempSlot -= skip;
+					row = tempSlot % accessoryPerColumn;
+					col = tempSlot / accessoryPerColumn;
+				}
+
+				yLoc = (int)((float)(num20) + (float)((row + 3) * 56) * Main.inventoryScale) + 4;
+				if (col > 0) {
+					xLoc = Main.screenWidth - 64 - 28 - 47 * 3 * col - 50;
+				}
+				else {
+					xLoc = Main.screenWidth - 64 - 28 - 47 * 3 * col;
+				}
+			}
+			
+			return true;
+		}
+
+		/// <summary>
+		/// Is run in AccessorySlotLoader.Draw. 
+		/// Generates a significant amount of functionality for the slot, despite being named drawing because vanilla.
+		/// At the end, calls this.DrawModded() where you can override to have custom drawing code for visuals.
+		/// Also includes creating hidevisibilitybutton.
+		/// </summary>
+		internal void DrawFunctional(Item[] access, bool[] visbility, int context, int slot, bool flag3, int xLoc, int yLoc) {
```

Comment:
Should use a `ref bool visibility` and drop the `slot` parameter

Reply:
DrawSlot() still needs a slot parameter, and propagates all the way to ItemSlot related functionality from this method.

So I've rewritten the method to just be drawing visibility so that the outcome makes sense.